server/src/main/java/io/crate/expression/symbol/GroupAndAggregateSemantics.java

```diff
@@ -76,6 +79,40 @@ public static void validate(List<Symbol> outputSymbols,
         }
     }
 
+    private static void validateGroupBySymbol(Symbol groupBy) {
+        groupBy.accept(SymbolTypeValidator.INSTANCE, null);
+    }
+
+    private static class SymbolTypeValidator extends SymbolVisitor<Void, Void> {
+
+        static final SymbolTypeValidator INSTANCE = new SymbolTypeValidator();
+
+        @Override
+        public Void visitSymbol(Symbol symbol, Void context) {
+            if (symbol.valueType() == DataTypes.UNDEFINED) {
+                throw new IllegalArgumentException(
+                    "Cannot group or aggregate on '" + symbol.toString() + "' efficiently with an undefined type." +
+                    " Use an explicit type cast."
+                );
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitLiteral(Literal symbol, Void context) {
+            if (symbol.valueType() == DataTypes.UNDEFINED && symbol.value() == null) {
+                // `NULL` is a valid case
+                return null;
+            }
+            return visitSymbol(symbol, context);
```

Comment:
```suggestion
            return null;
```

`visitSymbol` would do the `valueType() == UNDEFINED` check again.